WFLY-12321

fl4via opened the pull request:
Do not merge yet

Jira: https://issues.jboss.org/browse/WFLY-12321

This PR has to incorporate a few more feedbacks and will most likely be broken into more PRs and Jiras.
EAP Jira (temporary): https://issues.jboss.org/browse/JBEAP-17049

fl4via committed:
At several clustering infinispan classes, remove the G group identfier type @param comment

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/InfinispanBeanManager.java:
```diff
@@ -286,7 +286,6 @@ Node locatePrimaryOwner(I id) {
             InfinispanEjbLogger.ROOT_LOGGER.debugf("Could not find bean %s", id);
             return null;
         }
-        this.schedule(bean);
```
There needs to be be some action taken at the point to prevent concurrent expiration, in particular, when the bean is schedule to be expired by a different member than the one currently performing the invocation.

fl4via replied:
I added prepareRescheduling method below, btw, all methods and names are candidates for renaming if the name doesnt sound right, it is hard to get it right sometimes when you're not a native speaker ;-)

pferraro commented on clustering/api/src/main/java/org/wildfly/clustering/dispatcher/CommandDispatcher.java:
```diff
@@ -44,6 +44,14 @@
+     * @param member member node
+     * @return true if {@code member} is this local node
+     */
+    boolean isLocal(Node member);
```
We are removing this, correct?

fl4via replied:
+1 this is done

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/BeanExpirationScheduler.java:
```diff
@@ -38,46 +36,49 @@
+    private final BeanRemover<I, T> remover;
+    private final ExpirationConfiguration<T> expiration;
+    private final ExpirationTracker<I> expirationTracker;
+    private Future<?> expireTask;
```
expireTask should be volatile, no?

fl4via replied:
done

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/BeanExpirationScheduler.java:
```diff
@@ -38,46 +36,49 @@
+                Runnable task = new ExpirationTask();
+                // Make sure the expiration future map insertion happens before map removal (during task execution).
+                synchronized (this) {
+                    expireTask = this.expiration.getExecutor().schedule(task, timeout.toMillis(), TimeUnit.MILLISECONDS);
```
I think you want an additional null check here - since multiple concurrent threads can reach this synchronized block resulting in multiple expire task creation.
e.g.
```java
synchronized (this) {
   if (expireTask == null) {
       expireTask = ...;
   }
}
```

fl4via replied:
+1

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/BeanExpirationScheduler.java:
```diff
@@ -87,51 +88,59 @@ public void cancel(Locality locality) {
+                return;
+            }
+            task = expireTask;
+            expireTask = null;
```
Should we be concerned that the expireTask can be reset after this synchronized block completes?  Whose cancels the Future in that case?

fl4via replied:
I agree, I fixed that by checking if expireTask is different from null at the end of the file, before trying to reschedule it and reset the field.

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/InfinispanBeanManager.java:
```diff
@@ -229,8 +234,13 @@ private void cancel(Bean<I, T> bean) {
+            if (dispatcher.isLocal(node)) {
+                scheduler.schedule(bean.getId());
+                return;
+            }
```
```java
   if (this.dispatcherFactory.getGroup().isSingleton()) {
       scheduler.cancel(bean.getId());
       return;
   }
```

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/InfinispanBeanManager.java:
```diff
@@ -407,13 +417,17 @@ public T acquire() {
 
         @Override
         public boolean release() {
+            schedule(bean);
```
I don't think it makes sense to schedule expiration/passivation on release.  The last access time isn't updated until close(), and we already schedule for expiration/passivation then.

fl4via replied:
that's a great catch, I updated the PR with this

pferraro commented on ejb3/src/main/java/org/jboss/as/ejb3/cache/simple/SimpleCache.java:
```diff
@@ -169,6 +169,8 @@ public void release(V bean) {
                     remove(id);
                 }
+            } else {
+                remove(id);
```
This doesn't look right.  This would mean that a @Stateful EJB never retains its state across multiple invocations if it doesn't define a @StatefulTimeout.

fl4via replied:
I'll see, I'll skip this one

fl4via commented:
@pferraro I've updated the PR with all your feedback. Also, I removed the commit for "At LocalCommandDispatcher for clustering, executeOnMember should check if this node is equal to member, and not if it is equal to itself". I was thinking it shoudl have its own Jira, dont you agree? Or do you prefer to have the commit at this PR as it was before?

pferraro commented:
@fl4via 
> I've updated the PR with all your feedback. Also, I removed the commit for "At LocalCommandDispatcher for clustering, executeOnMember should check if this node is equal to member, and not if it is equal to itself". I was thinking it shoudl have its own Jira, dont you agree? Or do you prefer to have the commit at this PR as it was before?

Agreed. I merged it last week.
https://github.com/wildfly/wildfly/pull/12445

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/Scheduler.java:
```diff
@@ -35,6 +35,13 @@
+     * will be rescheduled again.
+     * @param id a bean identifier
+     */
+    void prepareRescheduling(I id);
```
Why create a new method?  Aren't we just changing the behavior of the existing cancel(...) method?

fl4via replied:
We are not, because cancel will invoke ExpirationTracker.forget, that will remove all entries from ExpirationTracker. If we keep calling cancel like before the gain is relatively small. I'm trying to make sure we schedule or cancel only once per call, I'm not sure if that's entirely feasible, but I'm trying to at least decrease the amount of times we add items to the tracker, to prevent the cost of invoking currentTimeInMillis

pferraro replied:
Ah, I see.  I assumed that there was no longer a use case for ExpirationTracker.forget, but there is.  So, cancel(I) is still triggered from cancel(Locaility, Locality) and from SchedulableBean.close() in the event that the bean is invalid (i.e. was removed).

fl4via commented on testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/dispatcher/bean/CommandDispatcherBean.java:
```diff
@@ -34,6 +34,11 @@ public void destroy() {
     }
 
+    @Override
+    public boolean isLocal(Node node) {
```
Somehow I must have failed to pull everything yesterday... I'll fix it, thanks for pointing this :-)

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/PrepareReeschedulingSchedulerCommand.java:
```diff
@@ -0,0 +1,48 @@
+ * the near future.
+ * @author Flavia Rainone
+ */
+public class PrepareReeschedulingSchedulerCommand<I> implements Command<Void, Scheduler<I>> {
```
Spelling fix: PrepareR**e**schedulingSchedulerCommand :)

fl4via replied:
Ooops, thanks :-)

rhusar commented:
With the latest changes `org.jboss.as.test.clustering.cluster.ejb.stateful.StatefulTimeoutTestCase#timeout` fails consistently.

fl4via commented:
Retest this please

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/BeanExpirationScheduler.java:
```diff
@@ -38,46 +36,57 @@
     @Override
     public void cancel(Locality locality) {
-        for (I id: this.expirationFutures.keySet()) {
+        for (I id: expirationTracker.getTrackedIds()) {
```
Iterating over this set will be prone to ConcurrentModificationExceptions.

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/EagerEvictionScheduler.java:
```diff
@@ -40,53 +40,66 @@
     @Override
     public void cancel(Locality locality) {
-        for (I id: this.evictionFutures.keySet()) {
+        for (I id: this.expirationTracker.getTrackedIds()) {
```
Iterating over this set will be prone to ConcurrentModificationExceptions.

stuartwdouglas commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/ExpirationTracker.java:
```diff
@@ -0,0 +1,253 @@
+                        firstExpiration != lastExpiration) {
+                    firstExpiration = firstExpiration.nextExpiration;
+                }
+                expiration.previousExpiration = null;
```
the nextExpriation must also me cleared, or an endless loop can result

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/ExpirationTracker.java:
```diff
@@ -0,0 +1,254 @@
+                firstExpiration = lastExpiration = expiration;
+            } else if (lastExpiration != expiration) {
+                // if the last expiration to occur is not current node, add it to the
+                // last of the expiration list
```
This logic makes the assumption that the newly tracked bean will be the last to expire.  While this is true when scheduling a bean for expiration at the end of an invocation, this is generally not the case when a member takes on the responsibility for expiring a given bean, i.e. when triggered from InfinispanBeanManager.schedule(Locality oldLocality, Locality newLocality).

fl4via replied:
Oh, I see. I'll try to think on a solution and I'll assert the perf gains with the benchmark to make sure this PR is really needed.

pferraro replied:
One solution is to replace the expiration tracker with a collection, e.g. io.undertow.util.ConcurrentDirectDeque.  We can then construct the scheduler with a queue implementation appropriate for the use case.  In the clustered case, where the assumption above is violated, we could use a priority queue, where elements are inserted according to their next expiration, rather then just appending to the end.  Insertions/removals would be slower log(N) instead of constant time, but that way we can still optimize for the local case.

fl4via committed:
[WFLY-12321] At InfinispanBeanManager, if node is singleton schedule a single task for eviction and another for bean expiration and never cancel that task; instead, update the expiration of the bean so it can be checked by the task if the bean has actually expired

fl4via commented:
@pferraro this commit is ready for review

fl4via commented:
Retest this please

fl4via commented:
Retest this please

fl4via commented:
Retest this please

bstansberry commented:
@fl4via @pferraro Is this meant for WF 18?

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/NonClusteredEagerEvictionScheduler.java:
```diff
@@ -0,0 +1,193 @@
+    public void evict(I id) {
+        try {
+            // Cache eviction is a local operation, so we need to broadcast this to the cluster
+            this.dispatcher.executeOnGroup(new EvictCommand<>(id));
```
Given that this is now a local-only scheduler, there is no need to dispatch an eviction command.

pferraro commented:
@bstansberry This is a performance optimization for local SFSBs.  Not essential for WF18, but still desirable.

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/NonClusteredEagerEvictionScheduler.java:
```diff
@@ -0,0 +1,188 @@
+        this.executor = executor;
+        this.idleTimeout = idleTimeout;
+        this.expirationTracker = new ExpirationTracker<>(idleTimeout);
+        this.dispatcher = dispatcherFactory.createCommandDispatcher(dispatcherName + "/eviction", evictor);
```
This scheduler no longer has any need to create a CommandDispatcher, no?

fl4via replied:
done

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/NonClusteredEagerEvictionScheduler.java:
```diff
@@ -0,0 +1,188 @@
+        // do nothing, this is a local-only scheduler
+    }
+
+    static class EvictCommand<I> implements Command<Void, BeanGroupEvictor<I>> {
```
I don't think this class is referenced any longer.

fl4via replied:
done

pferraro commented on clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/NonClusteredEagerEvictionScheduler.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Override
+    public void evict(I id) {
+        // do nothing, this is a local-only scheduler
```
This method needs to evict the bean, i.e. `this.evictor.evict(id);`

fl4via replied:
done

fl4via commented:
@pferraro  all requested changes are done

pferraro reviewed (approved):
Thanks Flavia. I think this is ready now.
